src/lib/core/datetime/date-adapter.ts

```diff
@@ -0,0 +1,170 @@
+   * @param locale The locale to use.
+   * @returns An ordered list of all date of the month names, starting with '1'.
+   */
+  abstract getDateNames(locale?: any): string[];
+
+  /**
+   * Gets a list of names for the weekdays.
+   * @param style The naming style (e.g. long = 'Sunday', short = 'Sun', narrow = 'S').
+   * @param locale The locale to use.
+   * @returns An ordered list of all weekday names, starting with Sunday.
+   */
+  abstract getDayNames(style: 'long' | 'short' | 'narrow', locale?: any): string[];
+
+  /**
+   * Gets the name for the year of the given date.
+   * @param date The date to get the year name for.
+   * @param locale The locale to use.
+   * @returns The name of the given year (e.g. '2017').
+   */
+  abstract getYearName(date: D, locale?: any): string;
+
+  /**
+   * Gets the name for the month and year of the given date.
+   * @param date The date to get the month and year name for.
+   * @param monthStyle The naming style for the month
+   *     (e.g. long = 'January', short = 'Jan', narrow = 'J').
+   * @param locale The locale to use.
+   * @returns The name of the month and year of the given date (e.g. 'Jan 2017').
+   */
+  abstract getMonthYearName(date: D, monthStyle: 'long' | 'short' | 'narrow', locale?: any): string;
+
+  /**
+   * Gets the first day of the week.
+   * @param locale The locale to use.
+   * @returns The first day of the week (0-indexed, 0 = Sunday).
+   */
+  abstract getFirstDayOfWeek(locale?: any): number;
+
+  /**
+   * Creates a date with the given year, month, and date.
+   * @param year The year of the date
+   * @param month The month of the date (0-indexed, 0 = January).
+   * @param date The date of month of the date.
+   * @returns The new date.
+   */
+  abstract create(year: number, month: number, date: number): D;
+
+  /**
+   * Gets today's date.
+   * @returns Today's date.
+   */
+  abstract today(): D;
+
+  /**
+   * Parses a date from a value.
+   * @param value The value to parse.
+   * @param fmt The format of the value.
+   * @param locale The locale of the value.
+   * @returns The parsed date, or null if date could not be parsed.
+   */
+  abstract parse(value: any, fmt?: any, locale?: any): D | null;
+
+  /**
+   * Formats a date as a string.
+   * @param date The value to parse.
+   * @param fmt The format to use for the result string.
+   * @param locale The locale to use for the result string.
+   * @returns The parsed date, or null if date could not be parsed.
+   */
+  abstract format(date: D, fmt?: any, locale?: any): string;
+
+  /**
+   * Adds the given number of years, months, and days to the given date.
+   * @param date The date to add to.
+   * @param amount The number of years, months, and days to add (may be negative).
+   * @returns A new date equal to the original with the given amount of time added.
+   */
+  abstract add(date: D, amount: {years?: number, months?: number, days?: number}): D;
```

Comment:
Hmm interesting... I assumed moment added years first then months then days, like how I do for `Date`, but it seems it does the opposite order. Do you think I should add largest unit first, smallest unit first, or only allow adding 1 unit at a time and force the user to determine the order?